core/shared_obj.h

```diff
@@ -0,0 +1,83 @@
+// Copyright (c) 2017, Nefeli Networks, Inc.
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+// * Redistributions of source code must retain the above copyright notice, this
+// list of conditions and the following disclaimer.
+//
+// * Redistributions in binary form must reproduce the above copyright notice,
+// this list of conditions and the following disclaimer in the documentation
+// and/or other materials provided with the distribution.
+//
+// * Neither the names of the copyright holders nor the names of their
+// contributors may be used to endorse or promote products derived from this
+// software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+#ifndef BESS_SHARED_OBJ_H_
+#define BESS_SHARED_OBJ_H_
+
+#include <functional>
+#include <memory>
+#include <mutex>
+#include <string>
+#include <typeindex>
+#include <typeinfo>
+#include <unordered_map>
+
+#include <glog/logging.h>
+
+namespace bess {
+
+using SharedObjectKey = std::pair<std::type_index, std::string>;
+
+struct PairHasher {
+  template <typename T1, typename T2>
+  std::size_t operator()(const std::pair<T1, T2> &p) const noexcept {
+    return std::hash<T1>{}(p.first) + std::hash<T2>{}(p.second);
+  }
+};
+
+extern std::unordered_map<SharedObjectKey, std::weak_ptr<void>, PairHasher>
+    shared_objects;
+
+template <typename T>
+std::shared_ptr<T> GetSharedObject(const std::string &name,
+                                   bool lookup_only = false) {
+  SharedObjectKey key = std::make_pair(std::type_index(typeid(T)), name);
+
+  static std::mutex mutex;
+  std::lock_guard<std::mutex> lock(mutex);
+  auto it = shared_objects.find(key);
+
+  if (it == shared_objects.end() || it->second.expired()) {
```

Comment:
I can't quite figure out why the unit tests are sometimes (?!) failing, but according to http://en.cppreference.com/w/cpp/memory/weak_ptr/lock the `lock` itself is atomic, so I think you can just use `.lock` here and test for failure, rather than an explicit `.expired` test...?

Reply:
The failure was due to the fact I used `unordered_map::insert()` as if it was `insert_or_assign()`. `insert()` doesn't overwrite an existing item, and some random reordering of gtest unittests triggered bug.

I think `lock()` doesn't need to be atomic here, since everything is protected by a giant mutex. Anyway your suggestion should make a much cleaner control flow. Updating the code...